### DISCUSSION TITLE
Question about specificying UID and GID for docker image; behavior changed in latest preview build?

### DISCUSSION THREAD

**Discussion by surfrock66** [photoprism/photoprism#2799] [General]
I just switched from the "latest" branch to the "preview" branch.  Before this, even though it's specified in the docker-config.yml to NOT do this, I had specified my UID something ~2000 (I have UID's 2000-3000 used for service accounts in my org/LDAP).  It worked, even though it was technically dis-advised.  It's not working now after switching to the preview branch (though I got it working by just not specifying the UID/GID and allowing it to run as root).

I could modify my photoprism user to be within UID 900-1200 as a one-off, but I was wondering what the reason was for that limit?  I've been googling to better understand how docker containers work and such and didn't see anything about that as a docker limitation (though my understanding is very early and rudimentary).

**Comment by lastzero** [comment]
See:
- https://github.com/photoprism/photoprism/pull/2730

**Comment by surfrock66** [comment]
Interesting; I can't say I fully understand it, but likely the upgrade from latest->preview came with the change from gosu to setpriv.  I went through the commit there (and the scripts it modified) but I'm not seeing where that boundary for the UID comes from; I know if I try to start it with UID 2022 now, the photoprism container gets this error:

`setpriv: uid 2022 not found, --init-groups requires an user that can be found on the system`

That's to be expected (and gosu probably worked around that and operated with UID's it didn't know), but I was wondering where the barrier for UID 1200 comes from.  Is that a limit in the base docker OS image?  Is there a reason for it to be confined to the normal range?

Also, is there any discussion about adding a similar option for the database container?  It seems it makes files on the docker host as uid 999 which is "systemd-coredump" on my ubuntu server.  

I'm spending so much time on this piece because I'm hoping to offload some of my storage onto my NAS, and I prefer precise UID/GID mapping across the org.  It's not a bug/issue/deal breaker by any measure, I'm just hoping to better understand.

**Comment by surfrock66** [comment]
By the way, I feel like it's tough to convey tone...I'm not complaining at all and am very appreciative for this project; these questions are for my understanding, I'm not in any way expecting my setup's esoteric properties to dictate any changes, I appreciate your time :)

**Comment by thequailman** [comment]
Just updated Photoprism and it keeps crashing on the setpriv piece:
```
setpriv: uid 2005 not found, --init-groups requires an user that can be found on the system
```

I'm setting a custom UID/GID that was working in the past:
```
-e PHOTOPRISM_UID=2005 -e PHOTOPRISM_GID=2005
```

Rolling back to 220901-bullseye fixes the issue.

**Reply by lastzero** [reply]
You are welcome to contribute improvements! As of now, the UID ranges 0, 33, 50-99, 500-600 and 900-1200 are supported, which is stated in our docker-compose.yml examples as well as in the official documentation:

- https://docs.photoprism.app/getting-started/config-options/#docker-image

**Comment by llitz** [comment]
I would suggest to get some inspiration from linuxserver.io and how they handle it.
IIRC, the first run checks the UID and GID env var (in photoprism's case it would be PHOTOPRISM_UID/GID) and see if the user exists in /etc/passwd and /etc/group, if it doesn't the user is created at that point.

In some of my personal containers, I have this:

```sh
PUID=${PUID:-1000}
PGID=${PGID:-1000}

USER=abc
GROUP=abc

if [ "$PUID" == "" -o "$PUID" == "0" ]; then
	USER=root
	GROUP=root
elif ! $(id abc&>/dev/null); then
	groupadd -g $PGID $USER
	useradd -m -u $PUID -g $GROUP $USER
fi
```

There's some room for improving the group portion (in case someone is only setting the group) but this worked for my use case without any issues (either run as root or as user configured in the environment).

**Reply by lastzero** [reply]
We had this but were running into very specific issues with Darktable and video transcoding as well as users who don't start the container as root.

**Reply by llitz** [reply]
Shouldn't it also cause issues for several of the scripts inside the scripts folder?

As a workaround, I have repurposed the davfs script to create the user and group I need.

**Reply by lastzero** [reply]
There should be no issues if you use the supported IDs.

**Reply by llitz** [reply]
While you are correct, this is not something that I can do.

I have had a server running for several years, and the user/group permissions are strict; I can't change them without affecting other services running on it. Using any of the pre-existing IDs, I don't have any access to the media.

If it is acceptable, one way to officially address it would be to have an extra option under PHOTOPRISM_INIT for custom_user and allow someone to provide a file "init-custom_user.sh" that would handle the user creation, which is similar to what I have now.

If having a custom_user is too specific, it would also be possible to specify a "custom_script, " allowing the end-user to run any prep routine they deem necessary.

**Comment by Stefan300381** [comment]
reviving this thread. I run photoprism on Synology and my original directory is owned by SynologyPhotos user with a UID > 100000. I know I can address this limitation with other methods, but I'd like to use the UID and GID settings.

**Reply by lastzero** [reply]
What UID specifically? Once we know the details, we can add support to our Docker image.

**Comment by vahidrahimibr** [comment]
Hello everyone
I need to add UID=1000 for debian 10 image's user 
how should I set the user ´śysadm´' UID=1000
